src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java

```diff
@@ -13,38 +13,110 @@
 // limitations under the License.
 package com.google.devtools.build.lib.exec;
 
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
 import com.github.luben.zstd.ZstdInputStream;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.devtools.build.lib.exec.Protos.ExecLogEntry;
 import com.google.devtools.build.lib.exec.Protos.File;
 import com.google.devtools.build.lib.exec.Protos.SpawnExec;
-import com.google.devtools.build.lib.util.Pair;
 import com.google.devtools.build.lib.util.io.MessageInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.ArrayDeque;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
 import java.util.Map;
 import java.util.SortedMap;
 import java.util.SortedSet;
 import java.util.TreeMap;
 import java.util.TreeSet;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.regex.MatchResult;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
 import javax.annotation.Nullable;
 
 /** Reconstructs an execution log in expanded format from the compact format representation. */
 public final class SpawnLogReconstructor implements MessageInputStream<SpawnExec> {
+  // The path of the repo mapping manifest file under the runfiles tree.
+  private static final String REPO_MAPPING_MANIFEST = "_repo_mapping";
+
+  // Examples:
+  // * bazel-out/k8-fastbuild/bin/pkg/file.txt (repo: null, path: "pkg/file.txt")
+  // * bazel-out/k8-fastbuild/bin/external/some_repo/pkg/file.txt (repo: "some_repo", path:
+  // "pkg/file.txt")
+  private static final Pattern DEFAULT_GENERATED_FILE_RUNFILES_PATH_PATTERN =
+      Pattern.compile("(?:bazel|blaze)-out/[^/]+/[^/]+/(?:external/(?<repo>[^/]+)/)?(?<path>.+)");
+
+  // Examples:
+  // * pkg/file.txt (repo: null, path: "pkg/file.txt")
+  // * external/some_repo/pkg/file.txt (repo: "some_repo", path: "pkg/file.txt")
+  private static final Pattern DEFAULT_SOURCE_FILE_RUNFILES_PATH_PATTERN =
+      Pattern.compile("(?:external/(?<repo>[^/]+)/)?(?<path>.+)");
```

Comment:
N.B. if the `external` prefix ends up being configurable as it's currently being discussed at https://github.com/bazelbuild/bazel/issues/23576#issuecomment-2357558975, we'd also have to include it in the compact log. cc @lberki for awareness.

Reply:
Since `--experimental_sibling_repository_layout` will probably be removed at some point, should I just remove support for it here and then we can add a new field? I can also keep the support around for as long as the flag is there.

Reply:
Let's keep `--experimental_sibling_repository_layout` supported for now.